src/Graph/NeighborSelection.cpp

```diff
@@ -1,115 +1,168 @@
 //
-// Created by weifang on 2/19/17.
+// Created by wei on 3/3/17.
 //
 
 #include "NeighborSelection.hpp"
 
 #include <Eigen/Dense>
-#include <stdexcept>
 #include <iostream>
-#include <unordered_map>
+#include <map>
 
 #ifdef BAZEL
-#include "../Models/ModelOptions.hpp"
+#include "Models/LinearRegression.hpp"
+#include "Models/Model.hpp"
 #else
-#include "ModelOptions.hpp"
+#include "../Models/LinearRegression.hpp"
+#include "../Models/Model.hpp"
 #endif
 
 using namespace Eigen;
 using namespace std;
 
-
-NeighborSelection::NeighborSelection() {
-    L1_reg = default_L1_reg;
-    L2_reg = default_L2_reg;
-    betaAll = MatrixXf::Ones(1,1);
-};
-
-
-NeighborSelection::NeighborSelection(const unordered_map<string, string>& options) {
+NeighborSelection::NeighborSelection(const unordered_map<string, string>& opts) {
     try {
-        L1_reg = stof(options.at("lambda"));
-    } catch (std::out_of_range& oor) {
-        L1_reg = default_L1_reg;
+        tolerance = stod(opts.at("tolerance"));
+    } catch(std::out_of_range& oor) {
+        tolerance = default_tolerance;
     }
     try {
-        L2_reg = stof(options.at("L2_lambda"));
-    } catch (std::out_of_range& oor) {
-        L2_reg = default_L2_reg;
+        learningRate = stod(opts.at("learning_rate"));
+    } catch(std::out_of_range& oor) {
+        learningRate = default_learning_rate;
     }
-    betaAll = MatrixXf::Ones(1,1);
-};
-
-void NeighborSelection::assertReadyToRun() {
-    // X should not be empty
-    const bool ready = ((X.rows() > 0) && (X.cols() > 0));
-    const string err_str = "X is empty.";
-    if (!ready) {
-        throw runtime_error(err_str);
+    try {
+        learningRate2 = stod(opts.at("learning_rate2"));
+    } catch(std::out_of_range& oor) {
+        learningRate2 = default_learning_rate2;
+    }
+    try {
+        innerStep1 = stoi(opts.at("innerStep1"));
+    } catch(std::out_of_range& oor) {
+        innerStep1 = default_inner_step1;
     }
+    try {
+        innerStep2 = stoi(opts.at("innerStep2"));
+    } catch(std::out_of_range& oor) {
+        innerStep2 = default_inner_step2;
+    }
+    prev_residue = numeric_limits<float>::max();
 }
 
-void NeighborSelection::setL1_reg(float l1) { L1_reg = l1; };
 
-void NeighborSelection::setL2_reg(float l2) { L2_reg = l2; };
+NeighborSelection::NeighborSelection() {
+    learningRate = default_learning_rate;
+    learningRate2 = default_learning_rate2;
```

Comment:
They are copied from ProximalGradientDescent. I've removed all unnecessary parameters from NeighborSelection in the latest commit.